File: codebase/eval/ingest.js
#!/usr/bin/env node
/**
 * ingest.js — Embed chunks và đẩy vào Qdrant
 *
 * Đã được cập nhật để dùng Jina embed (đồng bộ với embedder.js).
 * Nếu embeddings.json đã có sẵn thì dùng luôn, không embed lại.
 *
 * Cache format:
 *   { "T01-001": [0.012, -0.034, ...], ... }
 *
 * Biến môi trường:
 *   JINA_API_KEY (bắt buộc nếu embeddings.json chưa có)
 *   QDRANT_URL (mặc định http://localhost:6333)
 *
 * Chạy:
 *   node codebase/eval/ingest.js            # embed + upsert tất cả
 *   node codebase/eval/ingest.js --reset    # xoá collection trước khi upsert
 */

const fs = require("fs");
const path = require("path");
const https = require("https");
require("./loadenv.js");

const { ensureCollection, upsertPoints, count, deleteAll, COLLECTION } = require("./qdrant.js");

const CHUNKS = path.resolve(__dirname, "chunks.json");
const EMBEDDINGS = path.resolve(__dirname, "embeddings.json");
const MODEL_EMBED = "jina-embeddings-v3";
const DIM = 1024;
const BATCH = 32;

const apiKey = process.env.JINA_API_KEY;
if (!apiKey) {
  console.error("⚠ JINA_API_KEY chưa set — điền vào codebase/eval/.env");
  process.exit(1);
}

function sleep(ms) {
  return new Promise((r) => setTimeout(r, ms));
}

// ============== EMBED BATCH (Jina) ==============
function embedBatch(texts) {
  return new Promise((resolve, reject) => {
    const body = JSON.stringify({
      model: MODEL_EMBED,
      input: texts,
      dimensions: DIM,
      normalized: true,
      task: "retrieval.passage",
    });
    const req = https.request(
      {
        hostname: "api.jina.ai",
        path: "/v1/embeddings",
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          "Content-Length": Buffer.byteLength(body),
          Authorization: `Bearer ${apiKey}`,
        },
      },
      (res) => {
        let buf = "";
        res.on("data", (c) => (buf += c));
        res.on("end", () => {
          if (res.statusCode >= 400) {
            return reject(new Error(`embed HTTP ${res.statusCode}: ${buf.slice(0, 200)}`));
          }
          try {
            const j = JSON.parse(buf);
            return resolve(j.data.map((d) => d.embedding));
          } catch (e) {
            return reject(e);
          }
        });
      }
    );
    req.on("error", reject);
    req.write(body);
    req.end();
  });
}

function hashId(s) {
  let h = 5381;
  for (let i = 0; i < s.length; i++) {
    h = ((h << 5) + h + s.charCodeAt(i)) | 0;
  }
  return Math.abs(h);
}

async function main() {
  const shouldReset = process.argv.includes("--reset");
  if (shouldReset) {
    console.log("→ --reset: xoá collection trước...");
    try { await deleteAll(); } catch {}
  }

  console.log("→ Load chunks...");
  const chunks = JSON.parse(fs.readFileSync(CHUNKS, "utf-8"));

  // Build embeddings cache: load từ embeddings.json (từ embedder.js)
  let embeddings = {};
  if (fs.existsSync(EMBEDDINGS)) {
    const arr = JSON.parse(fs.readFileSync(EMBEDDINGS, "utf-8"));
    for (const e of arr) embeddings[e.code] = e.embedding;
    console.log(`→ Đã có ${Object.keys(embeddings).length} embeddings trong cache.`);
  }

  // Embed các chunks thiếu
  const missing = chunks.filter((c) => !embeddings[c.code]);
  if (missing.length > 0) {
    console.log(`→ Cần embed ${missing.length} đoạn mới...`);
    let done = 0;
    for (let i = 0; i < missing.length; i += BATCH) {
      const batch = missing.slice(i, i + BATCH);
      let attempt = 0;
      while (true) {
        try {
          const vecs = await embedBatch(batch.map((c) => c.text));
          for (let j = 0; j < batch.length; j++) {
            embeddings[batch[j].code] = vecs[j];
          }
          done += batch.length;
          console.log(`  ${done}/${missing.length}`);
          break;
        } catch (e) {
          attempt++;
          if (attempt >= 3) {
            console.error(`✗ Lỗi sau 3 lần: ${e.message}`);
            process.exit(1);
          }
          console.log(`  ! Retry ${attempt}/3: ${e.message.slice(0, 100)}`);
          await sleep(2000 * attempt);
        }
      }
      await sleep(200);
    }
    // Lưu embeddings.json để lần sau dùng lại
    const out = chunks.map((c) => ({ code: c.code, embedding: embeddings[c.code] })).filter((x) => x.embedding);
    fs.writeFileSync(EMBEDDINGS, JSON.stringify(out));
  }

  await ensureCollection();
  const before = await count();
  console.log(`→ Qdrant collection: ${before} vectors trước ingest`);

  // Build points
  const points = [];
  let missingCount = 0;
  for (const c of chunks) {
    const vec = embeddings[c.code];
    if (!vec) { missingCount++; continue; }
    points.push({
      id: hashId(c.code),
      vector: vec,
      payload: {
        code: c.code,
        source: c.source || "",
        file: c.file || "",
        text: c.text,
        charCount: c.charCount || c.text.length,
        ...(c.conversation_id ? { conversation_id: c.conversation_id } : {}),
        ...(c.kind ? { kind: c.kind } : {}),
      },
    });
  }

  if (missingCount > 0) console.log(`⚠ ${missingCount} chunks thiếu embedding, bỏ qua`);

  console.log(`→ Upsert ${points.length} points...`);
  await upsertPoints(points);

  const after = await count();
  console.log(`→ Qdrant sau ingest: ${after} vectors`);
  console.log("✓ Done.");
}

main().catch((e) => {
  console.error(e);
  process.exit(1);
});

File: codebase/eval_py/ingest.py
#!/usr/bin/env python3
"""
ingest.py — Embed chunks bằng Jina và đẩy vào Qdrant.

Port của codebase/eval/ingest.js.
Nếu embeddings.json đã có sẵn thì dùng luôn, không embed lại.

Sử dụng:
    python codebase/eval_py/ingest.py
    python codebase/eval_py/ingest.py --reset
"""

import asyncio
import json
import sys
import time
from pathlib import Path

import httpx

ROOT = Path(__file__).resolve().parents[1]
sys.path.insert(0, str(ROOT))

from vlearn import config  # noqa: E402 — sys.path set above
from vlearn.qdrant_client import ensure_collection, upsert_points, count, delete_all

CHUNKS_PATH = ROOT / "eval" / "chunks.json"
EMBEDDINGS_PATH = ROOT / "eval" / "embeddings.json"
MODEL_EMBED = "jina-embeddings-v3"
DIM = 1024
BATCH = 32


async def embed_batch(texts: list[str]) -> list[list[float]]:
    body = {"model": MODEL_EMBED, "input": texts, "dimensions": DIM, "normalized": True, "task": "retrieval.passage"}
    headers = {"Content-Type": "application/json", "Authorization": f"Bearer {config.JINA_API_KEY}"}
    async with httpx.AsyncClient(timeout=60) as client:
        response = await client.post("https://api.jina.ai/v1/embeddings", headers=headers, json=body)
    if response.status_code >= 400:
        raise RuntimeError(f"Jina embed HTTP {response.status_code}: {response.text[:200]}")
    return [item["embedding"] for item in response.json()["data"]]


def hash_id(s: str) -> int:
    h = 5381
    for ch in s.encode():
        h = ((h << 5) + h + ch) & 0xFFFFFFFF
    return h & 0x7FFFFFFF


async def main() -> None:
    should_reset = "--reset" in sys.argv
    if should_reset:
        print("→ --reset: xoá collection trước...")
        try:
            await delete_all()
        except Exception:
            pass

    print("→ Load chunks...")
    chunks = json.loads(CHUNKS_PATH.read_text(encoding="utf-8"))

    embeddings: dict[str, list[float]] = {}
    if EMBEDDINGS_PATH.exists():
        arr = json.loads(EMBEDDINGS_PATH.read_text(encoding="utf-8"))
        for entry in arr:
            embeddings[entry["code"]] = entry["embedding"]
        print(f"→ Đã có {len(embeddings)} embeddings trong cache.")

    missing = [c for c in chunks if c["code"] not in embeddings]
    if missing:
        if not config.JINA_API_KEY:
            print("⚠ JINA_API_KEY chưa set — không thể embed")
            sys.exit(1)
        print(f"→ Cần embed {len(missing)} đoạn mới...")
        done = 0
        for start in range(0, len(missing), BATCH):
            batch = missing[start:start + BATCH]
            for attempt in range(1, 4):
                try:
                    vecs = await embed_batch([c["text"] for c in batch])
                    for i, c in enumerate(batch):
                        embeddings[c["code"]] = vecs[i]
                    done += len(batch)
                    print(f"  {done}/{len(missing)}")
                    break
                except Exception as exc:
                    if attempt >= 3:
                        print(f"✗ Lỗi sau 3 lần: {exc}")
                        sys.exit(1)
                    print(f"  ! Retry {attempt}/3: {str(exc)[:100]}")
                    time.sleep(2 * attempt)
            time.sleep(0.2)
        out = [{"code": c["code"], "embedding": embeddings[c["code"]]} for c in chunks if c["code"] in embeddings]
        EMBEDDINGS_PATH.write_text(json.dumps(out), encoding="utf-8")

    await ensure_collection()
    before = await count()
    print(f"→ Qdrant collection: {before} vectors trước ingest")

    points = []
    missing_count = 0
    for c in chunks:
        vec = embeddings.get(c["code"])
        if not vec:
            missing_count += 1
            continue
        payload: dict = {"code": c["code"], "source": c.get("source", ""), "file": c.get("file", ""), "text": c["text"], "charCount": c.get("charCount", len(c["text"]))}
        if c.get("conversation_id"):
            payload["conversation_id"] = c["conversation_id"]
        if c.get("kind"):
            payload["kind"] = c["kind"]
        points.append({"id": hash_id(c["code"]), "vector": vec, "payload": payload})

    if missing_count:
        print(f"⚠ {missing_count} chunks thiếu embedding, bỏ qua")

    print(f"→ Upsert {len(points)} points...")
    await upsert_points(points)
    after = await count()
    print(f"→ Qdrant sau ingest: {after} vectors")
    print("✓ Done.")


if __name__ == "__main__":
    asyncio.run(main())
